mbedtls/src/ssl/async_io.rs

```diff
@@ -0,0 +1,253 @@
+                    #[cfg(debug_assertions)]
+                    if Self::digest(buf) == pending.digest {
+                        return Ok(buf);
+                    }
+
+                    #[cfg(not(debug_assertions))]
+                    return Ok(buf);
+                }
+                Err(IoError::new(
+                    IoErrorKind::Other,
+                    "mbedtls expects the same data if the previous call to poll_write() returned Poll::Pending",
+                ))
+            }
+        }
+    }
+
+    fn post_write(&mut self, buf: &[u8], res: &Poll<IoResult<usize>>) {
+        match res {
+            &Poll::Pending => {
+                if self.pending.is_none() {
+                    self.pending = Some(Box::new(DigestAndLen {
+                        #[cfg(debug_assertions)]
+                        digest: Self::digest(buf),
+                        len: buf.len(),
+                    }));
+                }
+            }
+            _ => {
+                self.pending = None;
+            }
+        }
+    }
+}
+
+/// Marker type for an IO implementation that implements both
+/// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+pub enum AsyncStream {}
+
+// TODO: Add enum `AnyAsyncIo` as marker type for an IO implementation that
+// doesn't implement `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`.
+
+// TODO: Add `AsyncIo` trait for async IO that that doesn't implement
+// `tokio::io::AsyncRead` and `tokio::io::AsyncWrite`. For example:
+//     pub trait AsyncIo {
+//        async fn recv(&mut self, buf: &mut [u8]) -> Result<usize>;
+//         async fn send(&mut self, buf: &[u8]) -> Result<usize>;
+//     }
+// Could implement by using `async-trait` crate or
+// #![feature(async_fn_in_trait)] or Associated Types
+
+impl<'a, 'b, 'c, IO: AsyncRead + AsyncWrite + std::marker::Unpin + 'static> IoCallback<AsyncStream>
+    for (&'a mut TaskContext<'b>, &'c mut IO)
+{
+    fn recv(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let mut buf = ReadBuf::new(buf);
+        let io = Pin::new(&mut self.1);
+        match io.poll_read(self.0, &mut buf) {
+            Poll::Ready(Ok(_)) => Ok(buf.filled().len()),
+            Poll::Ready(Err(_)) => Err(Error::NetRecvFailed),
+            Poll::Pending => Err(Error::SslWantRead),
+        }
+    }
+
+    fn send(&mut self, buf: &[u8]) -> Result<usize> {
+        let io = Pin::new(&mut self.1);
+        match io.poll_write(self.0, buf) {
+            Poll::Ready(Ok(n)) => Ok(n),
+            Poll::Ready(Err(_)) => Err(Error::NetSendFailed),
+            Poll::Pending => Err(Error::SslWantWrite),
+        }
+    }
+}
+
+impl<T: Unpin + AsyncRead + AsyncWrite + 'static> Context<T> {
+    pub async fn establish_async<IoType>(&mut self, io: T, hostname: Option<&str>) -> Result<()>
+    where
+        for<'c, 'cx> (&'c mut TaskContext<'cx>, &'c mut T): IoCallbackUnsafe<IoType>,
+    {
+        self.prepare_handshake(io, hostname)?;
```

Comment:
Nit: normally, function-local item definitions are placed at the top, so the code isn't interlaced with them.